execute/observation.go

```diff
@@ -259,70 +261,109 @@ func buildCombinedReports(
 	return combinedReports
 }
 
-// regroup converts the previous outcome to the observation format.
-// TODO: use same format for Observation and Outcome.
-func regroup(commitData []exectypes.CommitData) exectypes.CommitObservations {
-	groupedCommits := make(exectypes.CommitObservations)
-	for _, report := range commitData {
-		if _, ok := groupedCommits[report.SourceChain]; !ok {
-			groupedCommits[report.SourceChain] = []exectypes.CommitData{}
-		}
-		groupedCommits[report.SourceChain] = append(groupedCommits[report.SourceChain], report)
+// observeTokenDataForMessage observes token data for a given message.
+// It uses the token data observer to fetch token data
+// and handles errors by initializing the token data with an error.
+// It's okay to fetch 1 tokenData at a time without affecting peformance
+// Reason for that is we either use background observer that caches data and returns instantly, or even in case
+// we're using the sync observers, their implementations calls attestations also one by one.
+func (p *Plugin) observeTokenDataForMessage(
+	ctx context.Context,
+	lggr logger.Logger,
+	msg cciptypes.Message,
+) exectypes.MessageTokenData {
+	srcChain := msg.Header.SourceChainSelector
+	seqNum := msg.Header.SequenceNumber
+	msgObs := exectypes.MessageObservations{
+		srcChain: {
+			seqNum: msg,
+		},
 	}
-	return groupedCommits
+	msgTkData, err := p.tokenDataObserver.Observe(ctx, msgObs)
+	if err != nil {
+		lggr.Errorw("failed to observe token data", "err", err)
+		// In case of failure, initialize the token data with an error, that will prevent this specific token
+		// from being processed and sent later but won't stop the rest of messages
+		msgTkData = make(exectypes.TokenDataObservations)
+		msgTkData[srcChain] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+		msgTkData[srcChain][seqNum] = exectypes.NewMessageTokenData(exectypes.NewErrorTokenData(err))
+	}
+	return msgTkData[srcChain][seqNum]
 }
 
-func readAllMessages(
+// initSourceChainMaps initializes maps for a source chain if they don't exist
+func initSourceChainMaps(
+	srcChain cciptypes.ChainSelector,
+	messageObs exectypes.MessageObservations,
+	msgHashes exectypes.MessageHashes,
+	tkData exectypes.TokenDataObservations,
+) {
+	if _, ok := messageObs[srcChain]; !ok {
+		messageObs[srcChain] = make(map[cciptypes.SeqNum]cciptypes.Message)
+		msgHashes[srcChain] = make(map[cciptypes.SeqNum]cciptypes.Bytes32)
+		tkData[srcChain] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+	}
+}
```

Comment:
In terms of net value this function is 12 lines long but the body is just 5 lines and its only used in one place. Better to just get rid of it and put this code inline at the call site. Its also pretty clunky because it mutates inputs which is bad practice and leads to harder to understand code